orttraining/orttraining/python/orttraining_python_module.cc

```diff
@@ -0,0 +1,265 @@
+  auto initialize = [&]() {
+    static bool initialized = false;
+    if (initialized) {
+      return;
+    }
+    // Initialization of the module
+    InitArray();
+    Env::Default().GetTelemetryProvider().SetLanguageProjection(OrtLanguageProjection::ORT_PROJECTION_PYTHON);
+    ort_training_env = std::make_unique<ORTTrainingPythonEnv>();
+    initialized = true;
+  };
+  initialize();
+}
+
+ORTTrainingPythonEnv& GetTrainingEnv() {
+  if (!ort_training_env) {
+    InitializeTrainingEnv();
+  }
+  return *ort_training_env;
+}
+
+Environment& GetTrainingORTEnv() {
+  if (!ort_training_env) {
+    InitializeTrainingEnv();
+  }
+  return ort_training_env->GetORTEnv();
+}
+
+void ORTTrainingRegisterExecutionProviders(InferenceSession* sess, const std::vector<std::string>& provider_types,
+                                       const ProviderOptionsMap& provider_options_map) {
+  // search in environment
+  ORTTrainingPythonEnv& training_env = GetTrainingEnv();
+  for (auto provider_type : provider_types){
+    size_t hash;
+    if (GetProviderInstanceHash(provider_type, provider_options_map, hash)){
+      auto cached_provider_instance = training_env.GetExecutionProviderInstance(provider_type, hash);
+      if (!cached_provider_instance){
+        auto ep = CreateExecutionProviderInstance(sess, provider_type, provider_options_map);
+        if (ep){
+          training_env.AddExecutionProvider(provider_type, hash, std::move(ep));
+          cached_provider_instance = training_env.GetExecutionProviderInstance(provider_type, hash);
+        }
+      }
+      if (cached_provider_instance)
+        OrtPybindThrowIfError(sess->RegisterExecutionProvider(cached_provider_instance));
+    }
+    else{
+      // the EP doesn't support cache, register the instance to session
+      auto ep = CreateExecutionProviderInstance(sess, provider_type, provider_options_map);
+      if (ep)
+        OrtPybindThrowIfError(sess->RegisterExecutionProvider(std::move(ep)));
+    }
+  }
+}
+
+PYBIND11_MODULE(onnxruntime_pybind11_state, m) {
+  m.doc() = "pybind11 stateful interface to ORTTraining";
+  RegisterExceptions(m);
+  
+  Environment& env = GetTrainingORTEnv();
+  addGlobalMethods(m, env);
+  addObjectMethods(m, env, ORTTrainingRegisterExecutionProviders);
+  addOrtValueMethods(m);
+  addSparseTensorMethods(m);
+  addIoBindingMethods(m);
+
+#if !defined(__APPLE__) && \
+    (!defined(ORT_MINIMAL_BUILD) || defined(ORT_EXTENDED_MINIMAL_BUILD) || defined(ORT_MINIMAL_BUILD_CUSTOM_OPS))
+  Ort::SessionOptions tmp_options;
+  if (!InitProvidersSharedLibrary()) {
+    const logging::Logger& default_logger = logging::LoggingManager::DefaultLogger();
+    LOGS(default_logger, WARNING) << "Init provider bridge failed.";
+  }
+#endif
+  
+  addObjectMethodsForTraining(m, ORTTrainingRegisterExecutionProviders);
+#ifdef ENABLE_EAGER_MODE
+  addObjectMethodsForEager(m);
+#endif
+
+  // clean the ort training environment when python interpreter exit
```

Comment:
@souptc and I had a discussion regarding this. This fix doesn't work.  And we haven't find an easy fix yet.

So, the original problem is: 


Assume you have an exe and and DLL. The main executable has a global variable “a”:

static A a;

And the DLL has a global variable “b”:

static B b;

Then, 

|                                                                       |     Win     |     Linux    |  
|-----------------------------------------------------------------------|-------------|--------------|
|     Explicit   Linking without manual unloading                       |     ABAB    |     ABBA     |   
|     Explicit   Linking with manual unloading before “main” returns    |     ABBA    |     ABBA     |   
|     Implicit   Linking                                                |     BAAB    |     BAAB     |   

Now let's focus on the " Explicit   Linking " and "Linux" cases. The order is always ABBA.  
Then let's think A is ort env. B is a global variable in the cuda EP. So, by default B will be destructed before A.
Which means, in ort env's destructor, it shouldn't use any CUDA function, and should not do any clean up for any EP.
Now in our shared execution provider API, each EP should implement a shutdown function. But this is very very tricky! Think when you call into this function, you entered an area that all the surrounding variables are gone, what kind of cleanup you can do? I can't think of any.  That's why in  provider_bridge_ort.cc we have a comment saying:

"On Linux if we unload the cuda shared provider we crash" 

Surely it would crash!

However, this is just one side of the problem. The other side is about thread local variables. CUDA EP has thread local variables. A thread local variable is alive until:
1. The thread is gone
2. Or you explicitly cleaned the variable.

So, without training, we always cleanup the thread local CUDA contexts when the session ends. No EP lives longer than a session, and no session lives longer than ort env. So, when we destruct an ort env, all CUDA contexts should have already gone. But this PR changed the picture. 

If @souptc didn't put this fix, the env's destructor may need to call some cuda functions to clean up some cuda contexts that hold by the CUDA EPs. At that time, because the global variables in the CUDA EP DLL are all dead, the CUDA engine isn't working any more, you will got errors.  However, this fix has its own problem too. Python may call this atexit function while another ORT python object is still alive. Which means, ort env may get destroyed when an ort session is still alive! Then when we destruct the session, bad things happen.